### PR TITLE
rkbin: Update license to unfreeRedistributableFirmware

### PR DESCRIPTION
**Pull request by nakato**
## Description of changes

rkbin packages the firmware required for initialisation of rockchip SoC's.  There is a minor difference between unfreeRedistributable and unfreeRedistributableFirmware, in that nixpkgs considers the firmware variant "free" to allow it to be used to build images and channel releases.

This change will allow uboot to be built and cached for rockchip devices.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

#### Maintainer ping

@thefossguy

### CODE REVIEW

**Comment by thefossguy** [comment]
As per the repository [license](https://github.com/rockchip-linux/rkbin/blob/b4558da0860ca48bf1a571dd33ccba580b9abe23/LICENSE)

> Rockchip hereby grants to you a non-exclusive license (a) to use, copy,
distribute the Software; (b) to modify any source code as part of Software(if
any) and sublicense, distribute such modifications.
Except as expressively authorized by Rockchip in writing, you may NOT:
(a) decompile, reverse-engineer, dissemble, or attempt to derive any source
code from the Software;
(b) remove or obscure any copyright, patent, or trademark statement or notices
contained in the Software.


**I am not a lawyer** but this reads as, "_As long as you don't reverse engineer and remove any copyrights, feel free to distribute it_." I need an actual second opinion.

For good measure, can you modify the `installPhase` to copy the `LICENSE` to `$out/doc/LICENSE`? That way, we're not circumventing any license inclusion clause (if any is implied).

**Comment by nakato** [comment]
> **I am not a lawyer** but this reads as, "_As long as you don't reverse engineer and remove any copyrights, feel free to distribute it_." I need an actual second opinion.

I referenced raspberrypifw to determine why it doesn't require impure/unfree to build when included and its license is more strict than rockchips. Ref: Raspberry Pi [LICENCE.broadcom](https://github.com/raspberrypi/firmware/blob/master/boot/LICENCE.broadcom).

There are also have some uses of rkbin via *unfreeRedistributableFirmware* currently.
 https://github.com/NixOS/nixpkgs/blob/fedebce6732443f184a39a9536588feb46978206/pkgs/misc/uboot/default.nix#L280-L291


A more conservative approach to rkbin would have made since before Jul 26, 2023 as Rockchip did not provide a license with the blobs before that date, which made any disk image targeting rockchip with their blobs a bit of a grey area.

> For good measure, can you modify the `installPhase` to copy the `LICENSE` to `$out/doc/LICENSE`?

Good idea, especially with all the blobs added.

> That way, we're not circumventing any license inclusion clause (if any is implied).

**Comment by thefossguy** [comment]
> There are also have some uses of rkbin via _unfreeRedistributableFirmware_ currently.
> 
> https://github.com/NixOS/nixpkgs/blob/fedebce6732443f184a39a9536588feb46978206/pkgs/misc/uboot/default.nix#L280-L291


This basically means "download `rkbin` only for building `ubootNanoPCT4`. In layman terms, there is no redistribution of any binaries from the `rkbin` repository. But, when talking about the `rkbin` _package_ itself, that **will get redistributed**, which needs to have the correct license.

Also, a bit weird but this doesn't build on my aarch64 host:
```
$ nix build github:nixos/nixpkgs/fedebce6732443f184a39a9536588feb46978206#ubootNanoPCT4
       error: Package ‘arm-trusted-firmware-rk3399-2.10.0’ in /nix/store/sdxiik8s09k2h7dlnzaf8v7d2xzfzqab-source/pkgs/misc/arm-trusted-firmware/default.nix:82 has an unfree license (‘bsd3 unfreeRedistributable’), refusing to evaluate.

       a) To temporarily allow unfree packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNFREE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnfree = true; }
       in configuration.nix to override this.

       Alternatively you can configure a predicate to allow specific packages:
         { nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
             "arm-trusted-firmware-rk3399"
           ];
         }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnfree = true; }
       to ~/.config/nixpkgs/config.nix.
```

Does the same command build anything for you?

**Comment by nakato** [comment]
> > There are also have some uses of rkbin via _unfreeRedistributableFirmware_ currently.
> > https://github.com/NixOS/nixpkgs/blob/fedebce6732443f184a39a9536588feb46978206/pkgs/misc/uboot/default.nix#L280-L291
> 
> This basically means "download `rkbin` only for building `ubootNanoPCT4`. In layman terms, there is no redistribution of any binaries from the `rkbin` repository. But, when talking about the `rkbin` _package_ itself, that **will get redistributed**, which needs to have the correct license.

The blob for the targeted hardware gets embedded into the resulting u-boot binary, so that is redistribution.


> Also, a bit weird but this doesn't build on my aarch64 host:
> 
> ```
> $ nix build github:nixos/nixpkgs/fedebce6732443f184a39a9536588feb46978206#ubootNanoPCT4
>        error: Package ‘arm-trusted-firmware-rk3399-2.10.0’ in /nix/store/sdxiik8s09k2h7dlnzaf8v7d2xzfzqab-source/pkgs/misc/arm-trusted-firmware/default.nix:82 has an unfree license (‘bsd3 unfreeRedistributable’), refusing to evaluate.
> ```
> 
> Does the same command build anything for you?

No, it would fail too, as that is caused by arm-trusted-firmware-rk3399 being marked *unfreeRedistributabe*.

Following why that happened it comes down to the hdcp binary in ATF.  The stance comes down to binaries in a MIT licensed repository are not covered by MIT.  If that is the case then it should not be marked *unfreeRedistributabe* but as *unfree* as it has no license.  However, I'm guessing it is marked as redistributable because of intent.

**Comment by thefossguy** [comment]
LGTM then

**Comment by nakato** [comment]
@thefossguy Are you able to do the review and merge on this PR or is there someone else that needs pinged?

**Review by thefossguy — APPROVED** [review]

**Comment by thefossguy** [comment]
I don't have merge access. But @K900 can.

**Pull request merged**
